### ISSUE TITLE
Can Require Avoid Throwing `Uncaught Error: Invalid require call` For Shims?

### ISSUE THREAD

**Issue by machineghost** [requirejs/requirejs#1520]
Require recently gave me the following:

    require-dev.js:168 Uncaught Error: Invalid require call

Thankfully, [some random awesome person](https://alxndr.svbtle.com/requirejs-invalid-require-call-error) explained that that cryptic error came from me setting a shim dependency as a string, rather than as an array., If I hadn't found that explanation, I would have been completely mystified by the error (and would probably still be pouring through the Require source code/Stack Overflow trying to find an explanation).

Presumably the reason this error is so cryptic is because, at that point in the code, Require doesn't know enough to give a helpful error message.  However, there's no reason for the code to ever get to that point: the moment a user defines a shim, Require should be perfectly capable of checking whether that shim's dependency is a string or an array.  If it did it could then give a much more helpful error message (something to the effect of "you're an idiot and you defined your Foo shim with a string dependency").

Incidentally, at that point Require could also check whether `shims` is defined and `shim` isn't, and if so throw a helpful error (rather than letting the user think they configured `shim` when they really configured `shims`) ... but I suppose that's another ticket.

**Expected Behavior**
Require (politely) tells you you're an idiot for specifying the dependency as a string.

**Actual Behavior**
Require throws a cryptic error that doesn't even mention "shim", let alone tell you which one is broken or why.

**Reproduction Steps**

1) Configure Require with a shim that has a string dependency:

`require.config({shim:{foo: {deps: 'jquery'}}});`

2) Configure a module to depend on that shim:

`define(['foo'], () => {});`

3) Bring that module in to the Require environment:

`<script data-main="fooDependentModule" src="require.js"/>`

**Comment by jrburke** [comment]
Catching type errors very early is very helpful, and I see the usefulness of it. I think it is best to accomplish this with some static source checking, either something like an eslint plugin or maybe for people using a light typing system like TypeScript, something that can do that validation up front.

I do not plan to add runtime type checking of values passed to the APIs in the require.js file though, the size of that code for doing all the possible checks for all config values would noticeably increase the size of the library.

So I suggest a linting approach instead. Unfortunately I will not have the time to create such a node-based utility, but I can help someone else if they want to use the requirejs node package to do the scanning for API calls/config calls to then do type checking. [parse.findConfig](https://github.com/requirejs/r.js/blob/49b7a280e50ddeb5b150c4eeca356fb3a07ca1ac/build/jslib/parse.js#L436) in the internal r.js parse module in particular would be useful for this particular config type checking.

An approach like [xrayquire](https://github.com/requirejs/xrayquire) for inserting runtime adapters into the loader might also be a way to go, but I have not had the time to keep that up and to add API checks like this.

So I am closing as this is not something I would want to include in this repo's code. However, I do see the value of a general type checking tool that might run alongside this code, and I recognize the frustration of losing time to a problem like this.

**Comment by jsibre** [comment]
A year later, I just stumbled over this, and in my case machineghost was the random awesome person who'd written something helpful.  Before finding this thread, I think I spent 30-45 minutes chasing my tail on this exact issue.  

I can understand the concern around bloating the library with a lot of type checking.

How about adding a brief blurb to the [help page on requireargs](http://requirejs.org/docs/errors.html#requireargs) that informs people that shim configs can cause the problem, too?  

Something like: 

> Shims configured via require.config({...}) can also cause this problem.
`require.config({ 
  shim: { 
    'module': { 
      deps: 'dependency' 
    }
  }
}` 
Dependencies expressed via shims should use an array to list the dependencies.  
`require.config({ shim: { 'module': { deps: [ 'dependency' ] } } }`

Thanks!
